TestWatchWithProgressNotifyNoEvent: watch response expected in 3s, but timed out

via semaphore

```
=== RUN   TestWatchWithProgressNotifyNoEvent
2016-07-05 23:21:31.490108 I | integration: launching node5060386684892539366 (unix://localhost:node5060386684892539366.sock.bridge)
2016-07-05 23:21:31.490236 I | integration: launching node7331268175617031928 (unix://localhost:node7331268175617031928.sock.bridge)
2016-07-05 23:21:31.490647 I | integration: launching node5710343533434181243 (unix://localhost:node5710343533434181243.sock.bridge)
2016-07-05 23:21:31.511286 I | etcdserver: name = node7331268175617031928
2016-07-05 23:21:31.511347 I | etcdserver: data dir = /tmp/etcd185464527
2016-07-05 23:21:31.511382 I | etcdserver: name = node5060386684892539366
2016-07-05 23:21:31.511428 I | etcdserver: member dir = /tmp/etcd185464527/member
2016-07-05 23:21:31.511478 I | etcdserver: data dir = /tmp/etcd739442146
2016-07-05 23:21:31.511545 I | etcdserver: heartbeat = 10ms
2016-07-05 23:21:31.511588 I | etcdserver: election = 100ms
2016-07-05 23:21:31.511617 I | etcdserver: snapshot count = 0
2016-07-05 23:21:31.511659 I | etcdserver: advertise client URLs = unix://127.0.0.1:21698.29509.sock
2016-07-05 23:21:31.511696 I | etcdserver: initial advertise peer URLs = unix://127.0.0.1:21697.29509.sock
2016-07-05 23:21:31.511736 I | etcdserver: name = node5710343533434181243
2016-07-05 23:21:31.511776 I | etcdserver: data dir = /tmp/etcd584025561
2016-07-05 23:21:31.511801 I | etcdserver: member dir = /tmp/etcd584025561/member
2016-07-05 23:21:31.511823 I | etcdserver: heartbeat = 10ms
2016-07-05 23:21:31.511843 I | etcdserver: election = 100ms
2016-07-05 23:21:31.511859 I | etcdserver: snapshot count = 0
2016-07-05 23:21:31.511885 I | etcdserver: advertise client URLs = unix://127.0.0.1:21702.29509.sock
2016-07-05 23:21:31.511905 I | etcdserver: member dir = /tmp/etcd739442146/member
2016-07-05 23:21:31.511967 I | etcdserver: initial advertise peer URLs = unix://127.0.0.1:21701.29509.sock
2016-07-05 23:21:31.512079 I | etcdserver: initial cluster = node5060386684892539366=unix://127.0.0.1:21699.29509.sock,node5710343533434181243=unix://127.0.0.1:21701.29509.sock,node7331268175617031928=unix://127.0.0.1:21697.29509.sock
2016-07-05 23:21:31.512169 I | etcdserver: initial cluster = node5060386684892539366=unix://127.0.0.1:21699.29509.sock,node5710343533434181243=unix://127.0.0.1:21701.29509.sock,node7331268175617031928=unix://127.0.0.1:21697.29509.sock
2016-07-05 23:21:31.512228 I | etcdserver: heartbeat = 10ms
2016-07-05 23:21:31.512336 I | etcdserver: election = 100ms
2016-07-05 23:21:31.512428 I | etcdserver: snapshot count = 0
2016-07-05 23:21:31.512529 I | etcdserver: advertise client URLs = unix://127.0.0.1:21700.29509.sock
2016-07-05 23:21:31.512624 I | etcdserver: initial advertise peer URLs = unix://127.0.0.1:21699.29509.sock
2016-07-05 23:21:31.512812 I | etcdserver: initial cluster = node5060386684892539366=unix://127.0.0.1:21699.29509.sock,node5710343533434181243=unix://127.0.0.1:21701.29509.sock,node7331268175617031928=unix://127.0.0.1:21697.29509.sock
2016-07-05 23:21:31.522336 I | etcdserver: starting member 60205805d9a1e047 in cluster 76c27663b7b4d090
2016-07-05 23:21:31.522441 I | raft: 60205805d9a1e047 became follower at term 0
2016-07-05 23:21:31.522488 I | raft: newRaft 60205805d9a1e047 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2016-07-05 23:21:31.522517 I | raft: 60205805d9a1e047 became follower at term 1
2016-07-05 23:21:31.523940 I | rafthttp: starting peer 32ba9cde2b81c6b7...
2016-07-05 23:21:31.524211 I | rafthttp: started HTTP pipelining with peer 32ba9cde2b81c6b7
2016-07-05 23:21:31.525565 I | rafthttp: started streaming with peer 32ba9cde2b81c6b7 (writer)
2016-07-05 23:21:31.526076 I | etcdserver: starting member 32ba9cde2b81c6b7 in cluster 76c27663b7b4d090
2016-07-05 23:21:31.526217 I | raft: 32ba9cde2b81c6b7 became follower at term 0
2016-07-05 23:21:31.526294 I | raft: newRaft 32ba9cde2b81c6b7 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2016-07-05 23:21:31.526360 I | raft: 32ba9cde2b81c6b7 became follower at term 1
2016-07-05 23:21:31.526428 I | rafthttp: started streaming with peer 32ba9cde2b81c6b7 (writer)
2016-07-05 23:21:31.528440 I | rafthttp: started peer 32ba9cde2b81c6b7
2016-07-05 23:21:31.528661 I | rafthttp: added peer 32ba9cde2b81c6b7
2016-07-05 23:21:31.528730 I | rafthttp: starting peer bde37fb564a079e9...
2016-07-05 23:21:31.528845 I | rafthttp: started streaming with peer 32ba9cde2b81c6b7 (stream Message reader)
2016-07-05 23:21:31.529041 I | rafthttp: started HTTP pipelining with peer bde37fb564a079e9
2016-07-05 23:21:31.529118 I | rafthttp: starting peer 60205805d9a1e047...
2016-07-05 23:21:31.529797 I | rafthttp: started HTTP pipelining with peer 60205805d9a1e047
2016-07-05 23:21:31.529961 I | etcdserver: starting member bde37fb564a079e9 in cluster 76c27663b7b4d090
2016-07-05 23:21:31.530354 I | raft: bde37fb564a079e9 became follower at term 0
2016-07-05 23:21:31.530447 I | raft: newRaft bde37fb564a079e9 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2016-07-05 23:21:31.530500 I | raft: bde37fb564a079e9 became follower at term 1
2016-07-05 23:21:31.530553 I | rafthttp: started streaming with peer bde37fb564a079e9 (writer)
2016-07-05 23:21:31.530687 I | rafthttp: started streaming with peer 32ba9cde2b81c6b7 (stream MsgApp v2 reader)
2016-07-05 23:21:31.531951 I | rafthttp: started streaming with peer bde37fb564a079e9 (writer)
2016-07-05 23:21:31.532525 I | rafthttp: started streaming with peer 60205805d9a1e047 (writer)
2016-07-05 23:21:31.532755 I | rafthttp: started streaming with peer 60205805d9a1e047 (writer)
2016-07-05 23:21:31.533175 I | rafthttp: starting peer 32ba9cde2b81c6b7...
2016-07-05 23:21:31.533607 I | rafthttp: started HTTP pipelining with peer 32ba9cde2b81c6b7
2016-07-05 23:21:31.534906 I | rafthttp: started peer bde37fb564a079e9
2016-07-05 23:21:31.535173 I | rafthttp: started streaming with peer bde37fb564a079e9 (stream Message reader)
2016-07-05 23:21:31.535355 I | rafthttp: added peer bde37fb564a079e9
2016-07-05 23:21:31.538063 I | rafthttp: started streaming with peer 32ba9cde2b81c6b7 (writer)
2016-07-05 23:21:31.538461 I | rafthttp: started streaming with peer 60205805d9a1e047 (stream Message reader)
2016-07-05 23:21:31.539595 I | etcdserver: set snapshot count to default 10000
2016-07-05 23:21:31.539667 I | etcdserver: starting server... [version: 3.0.0+git, cluster version: to_be_decided]
2016-07-05 23:21:31.540638 I | rafthttp: started peer 60205805d9a1e047
2016-07-05 23:21:31.540904 I | rafthttp: added peer 60205805d9a1e047
2016-07-05 23:21:31.540994 I | rafthttp: starting peer bde37fb564a079e9...
2016-07-05 23:21:31.544331 I | rafthttp: started HTTP pipelining with peer bde37fb564a079e9
2016-07-05 23:21:31.544398 I | rafthttp: started streaming with peer 60205805d9a1e047 (stream MsgApp v2 reader)
2016-07-05 23:21:31.548201 I | rafthttp: started streaming with peer bde37fb564a079e9 (stream MsgApp v2 reader)
2016-07-05 23:21:31.548358 I | rafthttp: started streaming with peer 32ba9cde2b81c6b7 (writer)
2016-07-05 23:21:31.551463 I | rafthttp: started peer 32ba9cde2b81c6b7
2016-07-05 23:21:31.554038 I | rafthttp: added peer 32ba9cde2b81c6b7
2016-07-05 23:21:31.554216 I | raft: 60205805d9a1e047 is starting a new election at term 1
2016-07-05 23:21:31.554397 I | raft: 60205805d9a1e047 became candidate at term 2
2016-07-05 23:21:31.554501 I | raft: 60205805d9a1e047 received vote from 60205805d9a1e047 at term 2
2016-07-05 23:21:31.554570 I | raft: 60205805d9a1e047 [logterm: 1, index: 3] sent vote request to 32ba9cde2b81c6b7 at term 2
2016-07-05 23:21:31.554700 I | raft: 60205805d9a1e047 [logterm: 1, index: 3] sent vote request to bde37fb564a079e9 at term 2
2016-07-05 23:21:31.554803 I | integration: launched node5710343533434181243 (unix://localhost:node5710343533434181243.sock.bridge)
2016-07-05 23:21:31.554866 I | rafthttp: starting peer 60205805d9a1e047...
2016-07-05 23:21:31.555024 I | rafthttp: started streaming with peer 32ba9cde2b81c6b7 (stream MsgApp v2 reader)
2016-07-05 23:21:31.555160 I | rafthttp: started streaming with peer bde37fb564a079e9 (stream MsgApp v2 reader)
2016-07-05 23:21:31.555639 I | membership: added member 32ba9cde2b81c6b7 [unix://127.0.0.1:21697.29509.sock] to cluster 76c27663b7b4d090
2016-07-05 23:21:31.555721 I | rafthttp: peer 60205805d9a1e047 became active
2016-07-05 23:21:31.555779 I | rafthttp: established a TCP streaming connection with peer 60205805d9a1e047 (stream MsgApp v2 reader)
2016-07-05 23:21:31.555926 I | rafthttp: started streaming with peer bde37fb564a079e9 (writer)
2016-07-05 23:21:31.556091 I | rafthttp: started streaming with peer bde37fb564a079e9 (writer)
2016-07-05 23:21:31.556160 I | rafthttp: started peer bde37fb564a079e9
2016-07-05 23:21:31.556395 I | rafthttp: added peer bde37fb564a079e9
2016-07-05 23:21:31.556458 I | etcdserver: set snapshot count to default 10000
2016-07-05 23:21:31.556496 I | rafthttp: established a TCP streaming connection with peer 60205805d9a1e047 (stream Message reader)
2016-07-05 23:21:31.556577 I | rafthttp: started HTTP pipelining with peer 60205805d9a1e047
2016-07-05 23:21:31.556681 I | membership: added member 60205805d9a1e047 [unix://127.0.0.1:21701.29509.sock] to cluster 76c27663b7b4d090
2016-07-05 23:21:31.556840 I | rafthttp: peer 32ba9cde2b81c6b7 became active
2016-07-05 23:21:31.556896 I | rafthttp: started streaming with peer 32ba9cde2b81c6b7 (stream Message reader)
2016-07-05 23:21:31.556952 I | rafthttp: established a TCP streaming connection with peer 32ba9cde2b81c6b7 (stream Message writer)
2016-07-05 23:21:31.557346 I | rafthttp: started streaming with peer bde37fb564a079e9 (stream Message reader)
2016-07-05 23:21:31.557459 I | membership: added member bde37fb564a079e9 [unix://127.0.0.1:21699.29509.sock] to cluster 76c27663b7b4d090
2016-07-05 23:21:31.557933 I | raft: 32ba9cde2b81c6b7 [logterm: 1, index: 3, vote: 0] ignored vote from 60205805d9a1e047 [logterm: 1, index: 3] at term 1: lease is not expired (remaining ticks: 1)
2016-07-05 23:21:31.558248 I | etcdserver: starting server... [version: 3.0.0+git, cluster version: to_be_decided]
2016-07-05 23:21:31.558491 I | rafthttp: established a TCP streaming connection with peer 32ba9cde2b81c6b7 (stream MsgApp v2 writer)
2016-07-05 23:21:31.558580 I | rafthttp: started streaming with peer 60205805d9a1e047 (writer)
2016-07-05 23:21:31.559639 I | rafthttp: started streaming with peer 60205805d9a1e047 (writer)
2016-07-05 23:21:31.559714 I | membership: added member 32ba9cde2b81c6b7 [unix://127.0.0.1:21697.29509.sock] to cluster 76c27663b7b4d090
2016-07-05 23:21:31.560460 I | membership: added member 60205805d9a1e047 [unix://127.0.0.1:21701.29509.sock] to cluster 76c27663b7b4d090
2016-07-05 23:21:31.561096 I | membership: added member bde37fb564a079e9 [unix://127.0.0.1:21699.29509.sock] to cluster 76c27663b7b4d090
2016-07-05 23:21:31.561591 I | rafthttp: started peer 60205805d9a1e047
2016-07-05 23:21:31.561850 I | rafthttp: added peer 60205805d9a1e047
2016-07-05 23:21:31.561915 I | etcdserver: set snapshot count to default 10000
2016-07-05 23:21:31.561980 I | etcdserver: starting server... [version: 3.0.0+git, cluster version: to_be_decided]
2016-07-05 23:21:31.562368 I | rafthttp: established a TCP streaming connection with peer 60205805d9a1e047 (stream Message writer)
2016-07-05 23:21:31.562592 I | rafthttp: established a TCP streaming connection with peer 32ba9cde2b81c6b7 (stream Message reader)
2016-07-05 23:21:31.562904 I | rafthttp: started streaming with peer 60205805d9a1e047 (stream MsgApp v2 reader)
2016-07-05 23:21:31.563496 I | rafthttp: started streaming with peer 60205805d9a1e047 (stream Message reader)
2016-07-05 23:21:31.563791 I | membership: added member 32ba9cde2b81c6b7 [unix://127.0.0.1:21697.29509.sock] to cluster 76c27663b7b4d090
2016-07-05 23:21:31.564004 I | integration: launched node7331268175617031928 (unix://localhost:node7331268175617031928.sock.bridge)
2016-07-05 23:21:31.564347 I | membership: added member 60205805d9a1e047 [unix://127.0.0.1:21701.29509.sock] to cluster 76c27663b7b4d090
2016-07-05 23:21:31.564908 I | membership: added member bde37fb564a079e9 [unix://127.0.0.1:21699.29509.sock] to cluster 76c27663b7b4d090
2016-07-05 23:21:31.565060 I | rafthttp: peer bde37fb564a079e9 became active
2016-07-05 23:21:31.565125 I | rafthttp: established a TCP streaming connection with peer bde37fb564a079e9 (stream Message writer)
2016-07-05 23:21:31.565481 I | rafthttp: peer 60205805d9a1e047 became active
2016-07-05 23:21:31.565544 I | rafthttp: established a TCP streaming connection with peer 60205805d9a1e047 (stream Message reader)
2016-07-05 23:21:31.565755 I | rafthttp: peer bde37fb564a079e9 became active
2016-07-05 23:21:31.565819 I | rafthttp: established a TCP streaming connection with peer bde37fb564a079e9 (stream MsgApp v2 reader)
2016-07-05 23:21:31.566269 I | rafthttp: established a TCP streaming connection with peer bde37fb564a079e9 (stream MsgApp v2 writer)
2016-07-05 23:21:31.566668 I | rafthttp: peer 32ba9cde2b81c6b7 became active
2016-07-05 23:21:31.566789 I | rafthttp: established a TCP streaming connection with peer 32ba9cde2b81c6b7 (stream MsgApp v2 reader)
2016-07-05 23:21:31.566941 I | rafthttp: established a TCP streaming connection with peer 60205805d9a1e047 (stream MsgApp v2 reader)
2016-07-05 23:21:31.567771 I | rafthttp: established a TCP streaming connection with peer 60205805d9a1e047 (stream MsgApp v2 writer)
2016-07-05 23:21:31.570420 I | rafthttp: established a TCP streaming connection with peer 60205805d9a1e047 (stream Message writer)
2016-07-05 23:21:31.570492 I | rafthttp: established a TCP streaming connection with peer bde37fb564a079e9 (stream Message reader)
2016-07-05 23:21:31.570581 I | rafthttp: established a TCP streaming connection with peer 32ba9cde2b81c6b7 (stream MsgApp v2 reader)
2016-07-05 23:21:31.570830 I | rafthttp: established a TCP streaming connection with peer bde37fb564a079e9 (stream MsgApp v2 writer)
2016-07-05 23:21:31.570931 I | rafthttp: established a TCP streaming connection with peer 32ba9cde2b81c6b7 (stream Message writer)
2016-07-05 23:21:31.571020 I | raft: bde37fb564a079e9 [term: 1] received a MsgVote message with higher term from 60205805d9a1e047 [term: 2]
2016-07-05 23:21:31.571161 I | raft: bde37fb564a079e9 became follower at term 2
2016-07-05 23:21:31.571271 I | raft: bde37fb564a079e9 [logterm: 1, index: 3, vote: 0] voted for 60205805d9a1e047 [logterm: 1, index: 3] at term 2
2016-07-05 23:21:31.571421 I | rafthttp: established a TCP streaming connection with peer bde37fb564a079e9 (stream Message reader)
2016-07-05 23:21:31.571510 I | rafthttp: established a TCP streaming connection with peer 32ba9cde2b81c6b7 (stream MsgApp v2 writer)
2016-07-05 23:21:31.571565 I | rafthttp: established a TCP streaming connection with peer 60205805d9a1e047 (stream MsgApp v2 writer)
2016-07-05 23:21:31.571612 I | rafthttp: established a TCP streaming connection with peer bde37fb564a079e9 (stream Message writer)
2016-07-05 23:21:31.571663 I | rafthttp: established a TCP streaming connection with peer 32ba9cde2b81c6b7 (stream Message reader)
2016-07-05 23:21:31.571742 I | rafthttp: established a TCP streaming connection with peer bde37fb564a079e9 (stream MsgApp v2 reader)
2016-07-05 23:21:31.571891 I | raft: 60205805d9a1e047 received vote from bde37fb564a079e9 at term 2
2016-07-05 23:21:31.571985 I | raft: 60205805d9a1e047 [quorum:2] has received 2 votes and 0 vote rejections
2016-07-05 23:21:31.572101 I | raft: 60205805d9a1e047 became leader at term 2
2016-07-05 23:21:31.572196 I | raft: raft.node: 60205805d9a1e047 elected leader 60205805d9a1e047 at term 2
2016-07-05 23:21:31.572574 I | integration: launched node5060386684892539366 (unix://localhost:node5060386684892539366.sock.bridge)
2016-07-05 23:21:31.572892 I | raft: 32ba9cde2b81c6b7 [term: 1] received a MsgApp message with higher term from 60205805d9a1e047 [term: 2]
2016-07-05 23:21:31.572974 I | raft: 32ba9cde2b81c6b7 became follower at term 2
2016-07-05 23:21:31.573033 I | raft: raft.node: 32ba9cde2b81c6b7 elected leader 60205805d9a1e047 at term 2
2016-07-05 23:21:31.573249 I | raft: raft.node: bde37fb564a079e9 elected leader 60205805d9a1e047 at term 2
2016-07-05 23:21:31.577135 I | etcdserver: published {Name:node5710343533434181243 ClientURLs:[unix://127.0.0.1:21702.29509.sock]} to cluster 76c27663b7b4d090
2016-07-05 23:21:31.577565 I | etcdserver: published {Name:node7331268175617031928 ClientURLs:[unix://127.0.0.1:21698.29509.sock]} to cluster 76c27663b7b4d090
2016-07-05 23:21:31.577626 I | etcdserver: published {Name:node5060386684892539366 ClientURLs:[unix://127.0.0.1:21700.29509.sock]} to cluster 76c27663b7b4d090
2016-07-05 23:21:31.577686 I | etcdserver: setting up the initial cluster version to 3.0
2016-07-05 23:21:31.578706 N | membership: set the initial cluster version to 3.0
2016-07-05 23:21:31.578880 N | membership: set the initial cluster version to 3.0
2016-07-05 23:21:31.578981 N | membership: set the initial cluster version to 3.0
2016-07-05 23:21:40.577038 W | etcdserver: failed to send out heartbeat on time (exceeded the 10ms timeout for 2.140213ms)
2016-07-05 23:21:40.577118 W | etcdserver: server is likely overloaded
2016-07-05 23:21:40.577185 W | etcdserver: failed to send out heartbeat on time (exceeded the 10ms timeout for 2.249435ms)
2016-07-05 23:21:40.577229 W | etcdserver: server is likely overloaded
2016-07-05 23:21:40.592744 I | integration: terminating node7331268175617031928 (unix://localhost:node7331268175617031928.sock.bridge)
2016-07-05 23:21:40.594437 I | rafthttp: stopping peer 60205805d9a1e047...
2016-07-05 23:21:40.595429 I | rafthttp: closed the TCP streaming connection with peer 60205805d9a1e047 (stream MsgApp v2 writer)
2016-07-05 23:21:40.595524 I | rafthttp: stopped streaming with peer 60205805d9a1e047 (writer)
2016-07-05 23:21:40.596596 I | rafthttp: closed the TCP streaming connection with peer 60205805d9a1e047 (stream Message writer)
2016-07-05 23:21:40.596796 I | rafthttp: stopped streaming with peer 60205805d9a1e047 (writer)
2016-07-05 23:21:40.596990 W | rafthttp: lost the TCP streaming connection with peer 32ba9cde2b81c6b7 (stream MsgApp v2 reader)
2016-07-05 23:21:40.597062 I | rafthttp: stopped HTTP pipelining with peer 60205805d9a1e047
2016-07-05 23:21:40.597253 W | rafthttp: lost the TCP streaming connection with peer 60205805d9a1e047 (stream MsgApp v2 reader)
2016-07-05 23:21:40.597317 I | rafthttp: stopped streaming with peer 60205805d9a1e047 (stream MsgApp v2 reader)
2016-07-05 23:21:40.597479 W | rafthttp: lost the TCP streaming connection with peer 60205805d9a1e047 (stream Message reader)
2016-07-05 23:21:40.597541 I | rafthttp: stopped streaming with peer 60205805d9a1e047 (stream Message reader)
2016-07-05 23:21:40.597602 I | rafthttp: stopped peer 60205805d9a1e047
2016-07-05 23:21:40.597652 I | rafthttp: stopping peer bde37fb564a079e9...
2016-07-05 23:21:40.598662 I | rafthttp: closed the TCP streaming connection with peer bde37fb564a079e9 (stream MsgApp v2 writer)
2016-07-05 23:21:40.598729 I | rafthttp: stopped streaming with peer bde37fb564a079e9 (writer)
2016-07-05 23:21:40.598850 W | rafthttp: lost the TCP streaming connection with peer 32ba9cde2b81c6b7 (stream Message reader)
2016-07-05 23:21:40.599090 W | rafthttp: lost the TCP streaming connection with peer 32ba9cde2b81c6b7 (stream MsgApp v2 reader)
2016-07-05 23:21:40.599783 I | rafthttp: closed the TCP streaming connection with peer bde37fb564a079e9 (stream Message writer)
2016-07-05 23:21:40.599842 I | rafthttp: stopped streaming with peer bde37fb564a079e9 (writer)
2016-07-05 23:21:40.599959 I | rafthttp: stopped HTTP pipelining with peer bde37fb564a079e9
2016-07-05 23:21:40.600065 W | rafthttp: lost the TCP streaming connection with peer 32ba9cde2b81c6b7 (stream Message reader)
2016-07-05 23:21:40.600145 W | rafthttp: lost the TCP streaming connection with peer bde37fb564a079e9 (stream MsgApp v2 reader)
2016-07-05 23:21:40.600213 I | rafthttp: stopped streaming with peer bde37fb564a079e9 (stream MsgApp v2 reader)
2016-07-05 23:21:40.600412 W | rafthttp: lost the TCP streaming connection with peer bde37fb564a079e9 (stream Message reader)
2016-07-05 23:21:40.600480 E | rafthttp: failed to read bde37fb564a079e9 on stream Message (net/http: request canceled)
2016-07-05 23:21:40.600520 I | rafthttp: peer bde37fb564a079e9 became inactive
2016-07-05 23:21:40.600573 I | rafthttp: stopped streaming with peer bde37fb564a079e9 (stream Message reader)
2016-07-05 23:21:40.600625 I | rafthttp: stopped peer bde37fb564a079e9
2016-07-05 23:21:40.620286 I | integration: terminated node7331268175617031928 (unix://localhost:node7331268175617031928.sock.bridge)
2016-07-05 23:21:40.620993 I | integration: terminating node5060386684892539366 (unix://localhost:node5060386684892539366.sock.bridge)
2016-07-05 23:21:40.622074 I | rafthttp: stopping peer 32ba9cde2b81c6b7...
2016-07-05 23:21:40.622934 E | rafthttp: failed to write 32ba9cde2b81c6b7 on stream Message (write unix 127.0.0.1:21701.29509.sock->@: write: broken pipe)
2016-07-05 23:21:40.622992 I | rafthttp: peer 32ba9cde2b81c6b7 became inactive
2016-07-05 23:21:40.623852 W | rafthttp: lost the TCP streaming connection with peer 32ba9cde2b81c6b7 (stream Message writer)
2016-07-05 23:21:40.624494 W | rafthttp: failed to process raft message (raft: stopped)
2016-07-05 23:21:40.625020 I | rafthttp: closed the TCP streaming connection with peer 32ba9cde2b81c6b7 (stream MsgApp v2 writer)
2016-07-05 23:21:40.625102 I | rafthttp: stopped streaming with peer 32ba9cde2b81c6b7 (writer)
2016-07-05 23:21:40.626178 I | rafthttp: closed the TCP streaming connection with peer 32ba9cde2b81c6b7 (stream Message writer)
2016-07-05 23:21:40.626236 I | rafthttp: stopped streaming with peer 32ba9cde2b81c6b7 (writer)
2016-07-05 23:21:40.626407 I | rafthttp: stopped HTTP pipelining with peer 32ba9cde2b81c6b7
2016-07-05 23:21:40.626481 I | rafthttp: stopped streaming with peer 32ba9cde2b81c6b7 (stream MsgApp v2 reader)
2016-07-05 23:21:40.626551 I | rafthttp: stopped streaming with peer 32ba9cde2b81c6b7 (stream Message reader)
2016-07-05 23:21:40.626610 I | rafthttp: stopped peer 32ba9cde2b81c6b7
2016-07-05 23:21:40.626644 I | rafthttp: stopping peer 60205805d9a1e047...
2016-07-05 23:21:40.627654 I | rafthttp: closed the TCP streaming connection with peer 60205805d9a1e047 (stream MsgApp v2 writer)
2016-07-05 23:21:40.627710 I | rafthttp: stopped streaming with peer 60205805d9a1e047 (writer)
2016-07-05 23:21:40.628088 W | rafthttp: lost the TCP streaming connection with peer bde37fb564a079e9 (stream Message reader)
2016-07-05 23:21:40.628333 W | rafthttp: lost the TCP streaming connection with peer bde37fb564a079e9 (stream MsgApp v2 reader)
2016-07-05 23:21:40.628680 I | rafthttp: closed the TCP streaming connection with peer 60205805d9a1e047 (stream Message writer)
2016-07-05 23:21:40.628725 I | rafthttp: stopped streaming with peer 60205805d9a1e047 (writer)
2016-07-05 23:21:40.628822 I | rafthttp: stopped HTTP pipelining with peer 60205805d9a1e047
2016-07-05 23:21:40.628985 W | rafthttp: lost the TCP streaming connection with peer 60205805d9a1e047 (stream MsgApp v2 reader)
2016-07-05 23:21:40.629035 E | rafthttp: failed to read 60205805d9a1e047 on stream MsgApp v2 (net/http: request canceled)
2016-07-05 23:21:40.629062 I | rafthttp: peer 60205805d9a1e047 became inactive
2016-07-05 23:21:40.629101 I | rafthttp: stopped streaming with peer 60205805d9a1e047 (stream MsgApp v2 reader)
2016-07-05 23:21:40.629237 W | rafthttp: lost the TCP streaming connection with peer 60205805d9a1e047 (stream Message reader)
2016-07-05 23:21:40.629299 I | rafthttp: stopped streaming with peer 60205805d9a1e047 (stream Message reader)
2016-07-05 23:21:40.629336 I | rafthttp: stopped peer 60205805d9a1e047
2016-07-05 23:21:40.646309 I | integration: terminated node5060386684892539366 (unix://localhost:node5060386684892539366.sock.bridge)
2016-07-05 23:21:40.646392 I | integration: terminating node5710343533434181243 (unix://localhost:node5710343533434181243.sock.bridge)
2016-07-05 23:21:40.646731 I | rafthttp: stopping peer 32ba9cde2b81c6b7...
2016-07-05 23:21:40.647771 I | rafthttp: closed the TCP streaming connection with peer 32ba9cde2b81c6b7 (stream MsgApp v2 writer)
2016-07-05 23:21:40.647841 I | rafthttp: stopped streaming with peer 32ba9cde2b81c6b7 (writer)
2016-07-05 23:21:40.648036 I | rafthttp: stopped streaming with peer 32ba9cde2b81c6b7 (writer)
2016-07-05 23:21:40.648108 I | rafthttp: stopped HTTP pipelining with peer 32ba9cde2b81c6b7
2016-07-05 23:21:40.648178 I | rafthttp: stopped streaming with peer 32ba9cde2b81c6b7 (stream MsgApp v2 reader)
2016-07-05 23:21:40.648249 I | rafthttp: stopped streaming with peer 32ba9cde2b81c6b7 (stream Message reader)
2016-07-05 23:21:40.648313 I | rafthttp: stopped peer 32ba9cde2b81c6b7
2016-07-05 23:21:40.648345 I | rafthttp: stopping peer bde37fb564a079e9...
2016-07-05 23:21:40.649258 I | rafthttp: closed the TCP streaming connection with peer bde37fb564a079e9 (stream MsgApp v2 writer)
2016-07-05 23:21:40.649302 I | rafthttp: stopped streaming with peer bde37fb564a079e9 (writer)
2016-07-05 23:21:40.650264 I | rafthttp: closed the TCP streaming connection with peer bde37fb564a079e9 (stream Message writer)
2016-07-05 23:21:40.650311 I | rafthttp: stopped streaming with peer bde37fb564a079e9 (writer)
2016-07-05 23:21:40.650696 I | rafthttp: stopped HTTP pipelining with peer bde37fb564a079e9
2016-07-05 23:21:40.650821 I | rafthttp: stopped streaming with peer bde37fb564a079e9 (stream MsgApp v2 reader)
2016-07-05 23:21:40.650942 I | rafthttp: stopped streaming with peer bde37fb564a079e9 (stream Message reader)
2016-07-05 23:21:40.651020 I | rafthttp: stopped peer bde37fb564a079e9
2016-07-05 23:21:40.667589 I | integration: terminated node5710343533434181243 (unix://localhost:node5710343533434181243.sock.bridge)
--- FAIL: TestWatchWithProgressNotifyNoEvent (9.18s)
	watch_test.go:491: watch response expected in 3s, but timed out
```

--- a/clientv3/integration/watch_test.go
+++ b/clientv3/integration/watch_test.go
@@ -487,7 +487,7 @@ func testWatchWithProgressNotify(t *testing.T, watchOnPut bool) {
 		} else if len(resp.Events) != 0 { // wait for notification otherwise
 			t.Fatalf("expected no events, but got %+v", resp.Events)
 		}
-	case <-time.After(2 * pi):
+	case <-time.After(time.Duration(1.5 * float64(pi))):
 		t.Fatalf("watch response expected in %v, but timed out", pi)
 	}
 }

--- a/etcdserver/api/v3rpc/watch.go
+++ b/etcdserver/api/v3rpc/watch.go
@@ -295,7 +295,8 @@ func (sws *serverWatchStream) sendLoop() {
 			}
 
 			sws.mu.Lock()
-			if _, ok := sws.progress[wresp.WatchID]; ok {
+			if len(evs) > 0 && sws.progress[wresp.WatchID] {
+				// elide next progress update if sent a key update
 				sws.progress[wresp.WatchID] = false
 			}
 			sws.mu.Unlock()

--- a/integration/v3_watch_test.go
+++ b/integration/v3_watch_test.go
@@ -971,7 +971,7 @@ func TestWatchWithProgressNotify(t *testing.T) {
 	}
 
 	// no more notification
-	rok, resp := waitResponse(wStream, testInterval+time.Second)
+	rok, resp := waitResponse(wStream, time.Second)
 	if !rok {
 		t.Errorf("unexpected pb.WatchResponse is received %+v", resp)
 	}

